docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
```

Comment:
I like explicit link to swagger as well - makes it easier for clients.  And
we could have the server generate it if we need to in the future by
introducing a new field on APIGroupLocation

On Mon, Nov 16, 2015 at 1:38 PM, Daniel Smith <notifications@github.com>
wrote:

> In docs/proposals/third-party-api-servers.md
> <https://github.com/kubernetes/kubernetes/pull/17191#discussion_r44962713>
> :
>
> > +
> > +  ObjectMeta
> > +
> > +  Spec ThirdPartyAPIServerSpec
> > +
> > +  Status ThirdPartyAPIServerStatus
> > +}
> > +
> > +// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
> > +// resource.
> > +type ThirdPartyAPIServerSpec struct {
> > +  // Base path of the API server.
> > +  BasePath string
> > +
> > +  // Path to the swagger spec.
> > +  SwaggerSpecPath string
>
> @nikhiljindal <https://github.com/nikhiljindal> @smarterclayton
> <https://github.com/smarterclayton> Below is how I am imagining it will
> work:
>
> The "main" apiserver serves a page at "/apis/", which has an entry for the
> group registered by this object. That entry has a URL to the delegate
> apiserver (e.g. http://delegate.apiserver.local/path/to/group). The page
> served there should have a link to the swagger for the group (or a list of
> versions, each with a link to their own swagger, if we break it up that
> much).
>
> When I wrote my above comment I was thinking that there'd be a mechanical
> transform from the group path to the swagger path, but now I think that
> it's better to explicitly link to the swagger from the group. Either way,
> it doesn't seem necessary to register the swagger page in advance? The main
> apiserver can always run discovery itself if it wants to find this
> information.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17191/files#r44962713>.
>